### PR TITLE
feat: display multifile projects

### LINKED ISSUE #44979: Make sure finished projects render on certs / timeline?
Nice to have:

- [ ] A user can open previews in full-screen in a new tab.

### PR DESCRIPTION
**Pull request by ojeytonwilliams**
Checklist:

- [x] I have read [freeCodeCamp's contribution guidelines](https://contribute.freecodecamp.org).
- [x] My pull request has a descriptive title (not a vague title like `Update index.md`)
- [x] My pull request targets the `main` branch of freeCodeCamp.
- [x] I have tested these changes either locally on my machine, or GitPod.

Closes https://github.com/freeCodeCamp/freeCodeCamp/issues/44979

To make QA easier, I've left in a commit that converts a project page to use the multi-file editor.  Though, I'll take it out if CI fails.

The commit messages (should) explain what's going on, but let me know if there's anything that doesn't make sense.

Edit: turns out it was breaking the CI.

### CODE REVIEW

**Review by moT01 — COMMENTED** [review]
This is looking real good @ojeytonwilliams 🎉 Not sure if this is even ready for review - you may still be working on some of the things I bring up here.

I looked over it pretty good. With the exception of my one concern in `/server/boot/challenge.js`, everything seems to be functionally working. Everything else I noticed is cosmetic...

#### On all the modals:

Pretty minor - but when viewing the code, the modal title is left aligned, when viewing the rendered preview, the title is centered. The JS code solutions have the title left aligned in the modal. It might be nice if they were all either one or the other. I would lean towards centered. I'm fine leaving this, though.

#### Settings and Timeline page:

The `View` button for new code/project viewing is not full width. All the other buttons are full width:
Screen Shot 2022-02-23 at 12 39 00 PM
^ screen shot of timeline. Top button is view -> code/project (not full width), bottom button is view -> frontend/backend. It's significantly more pronounced on the settings page.

#### Settings page:
Pretty minor - I think `View` is fine for the timeline, but it looks a little out of place on the settings page. I would change it to `Show solution` like we use for the projects with a URL and Github URL as the solution

#### Timeline:
For some reason, the modal that shows the code is more narrow on this page - not on the others. Probably nice to make it the same width.

#### Certificate page:
Probably shouldn't use the view button on the certifications:

Screen Shot 2022-02-23 at 12 14 16 PM
^ landing page has a URL solution - others have the multifile editor solution

Here's what the URL + github URL solutions look like:
Screen Shot 2022-02-23 at 12 22 35 PM

I would go with two text buttons/links instead of the view button - probably code/preview or something.
<hr>
That's all I got. Let me know what you're thinking @ojeytonwilliams - maybe I can take a look at some of these.

File: api-server/src/server/boot/challenge.js [hunk 1]
```diff
@@ -103,7 +103,7 @@ export function buildUserUpdate(
       )
     };
   } else {
-    completedChallenge = omit(_completedChallenge, ['files']);
+    completedChallenge = omit(_completedChallenge, ['files', 'challengeType']);
```
[the remark this answers is not in the document]
Comment (ojeytonwilliams) [reply api-server/src/server/boot/challenge.js:106]: Yeah, that's exactly what I was trying to do. Thanks for catching the shoddy logic!

**Comment by ojeytonwilliams** [comment]
@moT01 thanks for the detailed review.  I agree that the UI is pretty bad, for all the reasons you went into.  However, if it's okay with everyone, I'd like to get this in as is (after I've fixed the api logic).  Reason being, this PR already does quite a lot and the bad UI will only be visible once the RWD projects have been converted.

File: api-server/src/server/boot/challenge.js [hunk 2]
```diff
@@ -239,17 +239,22 @@ export function modernChallengeCompleted(req, res, next) {
...
         completedDate
       };
 
       if (challengeType === 14) {
-        data.isManuallyApproved = false;
+        completedChallenge.isManuallyApproved = false;
       }
 
-      const { alreadyCompleted, updateData } = buildUserUpdate(user, id, data);
+      const { alreadyCompleted, updateData } = buildUserUpdate(
```
[the remark this answers is not in the document]
Comment (ojeytonwilliams) [reply api-server/src/server/boot/challenge.js:261]: Yeah, well, 'data' is a bit generic!

**Review by moT01 — DISMISSED** [review]
> I'd like to get this in as is

I'm fine with that - LGTM 🎉

File: client/src/components/SolutionViewer/SolutionViewer.tsx [hunk 3]
```diff
@@ -1,66 +1,47 @@
...
-        ))
-      ) : (
-        <Panel
-          bsStyle='primary'
-          className='solution-viewer'
-          key={solution.slice(0, 10)}
-        >
-          <Panel.Heading>JS</Panel.Heading>
+      {solutions.map(({ fileKey, ext, contents }) => (
+        <Panel bsStyle='primary' className='solution-viewer' key={fileKey}>
```
Comment (ShaunSHamilton) [inline client/src/components/SolutionViewer/SolutionViewer.tsx:28]:
```suggestion
      {solutions.map(({ ext, contents }, i) => (
        <Panel bsStyle='primary' className='solution-viewer' key={i}>
```

**Review by ShaunSHamilton — CHANGES_REQUESTED** [review]
Other than small change, because `fileKey` is not actually unique, I am going to miss this:

```
<Modal.Header className='this-one?'
```

File: client/src/components/SolutionViewer/SolutionViewer.tsx (hunk 3 above)
Comment (nhcarrigan) [reply client/src/components/SolutionViewer/SolutionViewer.tsx:28]: Isn't it bad form to use the index of an array as the key?

Comment (ShaunSHamilton) [reply client/src/components/SolutionViewer/SolutionViewer.tsx:28]: I have heard that, but,  until someone explains why, I am going to keep doing it to things I am not popping.

Comment (ojeytonwilliams) [reply client/src/components/SolutionViewer/SolutionViewer.tsx:28]: It's explained here: https://robinpokorny.medium.com/index-as-a-key-is-an-anti-pattern-e0349aece318

TL;DR: React will assume that components with the same keys are the same components. This can be a problem if you re-order components in a list.

We're not re-ordering them, but it's a good practice in case someone decides to.

Anyways, the issue was that `fileKey` was missing, rather than not unique.  I've fixed that now, so the warnings should go away.

**Review by ShaunSHamilton — DISMISSED** [review]
Sure, LGTM 👍

**Review by moT01 — CHANGES_REQUESTED** [review]

File: client/src/templates/Challenges/redux/completion-epic.js [hunk 4]
```diff
@@ -35,9 +35,14 @@ function postChallenge(update, username) {
   const saveChallenge = postUpdate$(update).pipe(
     retry(3),
     switchMap(({ points }) => {
+      // TODO: do this all in ajax.ts
       const payloadWithClientProperties = {
         ...omit(update.payload, ['files']),
-        challengeFiles: update.payload.files ?? null
+        challengeFiles:
+          update.payload.files.map(({ key, ...rest }) => ({
```
[the remark this answers is not in the document]
Comment (ojeytonwilliams) [reply client/src/templates/Challenges/redux/completion-epic.js:42]: So, this broken every other challenge and no tests caught it.  Nice QAing!

**Review by ShaunSHamilton — APPROVED** [review]
This LTMG 👍

**Review by ahmaxed — APPROVED** [review]

**Review by moT01 — APPROVED** [review]
